url72c32f3c54_Quickhull3d_quickhull3d_tgz-pJ8-com_github_quickhull3d_SimpleExampleJ8/src/com/github/quickhull3d/QuickHull3D.java

```diff
@@ -0,0 +1,1334 @@
+    // }
+    // }
+
+    protected void initBuffers(int nump) {
+        if (pointBuffer.length < nump) {
+            Vertex[] newBuffer = new Vertex[nump];
+            vertexPointIndices = new int[nump];
+            for (int i = 0; i < pointBuffer.length; i++) {
+                newBuffer[i] = pointBuffer[i];
+            }
+            for (int i = pointBuffer.length; i < nump; i++) {
+                newBuffer[i] = new Vertex();
+            }
+            pointBuffer = newBuffer;
+        }
+        faces.clear();
+        claimed.clear();
+        numFaces = 0;
+        numPoints = nump;
+    }
+
+    protected void setPoints(double[] coords, int nump) {
+        for (int i = 0; i < nump; i++) {
+            Vertex vtx = pointBuffer[i];
+            vtx.pnt.set(coords[i * 3 + 0], coords[i * 3 + 1], coords[i * 3 + 2]);
+            vtx.index = i;
+        }
+    }
+
+    protected void setPoints(Point3d[] pnts, int nump) {
+        for (int i = 0; i < nump; i++) {
+            Vertex vtx = pointBuffer[i];
+            vtx.pnt.set(pnts[i]);
+            vtx.index = i;
+        }
+    }
+
+    protected void computeMaxAndMin() {
+        Vector3d max = new Vector3d();
+        Vector3d min = new Vector3d();
+
+        for (int i = 0; i < 3; i++) {
+            maxVtxs[i] = minVtxs[i] = pointBuffer[0];
+        }
+        max.set(pointBuffer[0].pnt);
+        min.set(pointBuffer[0].pnt);
+
+        for (int i = 1; i < numPoints; i++) {
+            Point3d pnt = pointBuffer[i].pnt;
+            if (pnt.x > max.x) {
+                max.x = pnt.x;
+                maxVtxs[0] = pointBuffer[i];
+            } else if (pnt.x < min.x) {
+                min.x = pnt.x;
+                minVtxs[0] = pointBuffer[i];
+            }
+            if (pnt.y > max.y) {
+                max.y = pnt.y;
+                maxVtxs[1] = pointBuffer[i];
+            } else if (pnt.y < min.y) {
+                min.y = pnt.y;
+                minVtxs[1] = pointBuffer[i];
+            }
+            if (pnt.z > max.z) {
+                max.z = pnt.z;
+                maxVtxs[2] = pointBuffer[i];
+            } else if (pnt.z < min.z) {
+                min.z = pnt.z;
+                minVtxs[2] = pointBuffer[i];
+            }
+        }
+
+        // this epsilon formula comes from QuickHull, and I'm
+        // not about to quibble.
+        charLength = Math.max(max.x - min.x, max.y - min.y);
+        charLength = Math.max(max.z - min.z, charLength);
+        if (explicitTolerance == AUTOMATIC_TOLERANCE) {
+            tolerance =
+                    3 * DOUBLE_PREC * (Math.max(Math.abs(max.x), Math.abs(min.x)) + Math.max(Math.abs(max.y), Math.abs(min.y)) + Math.max(Math.abs(max.z), Math.abs(min.z)));
+        } else {
+            tolerance = explicitTolerance;
+        }
+    }
+
+    /**
+     * Creates the initial simplex from which the hull will be built.
+     */
+    protected void createInitialSimplex() throws IllegalArgumentException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 28. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 658-744 into a separate method.